### ISSUE TITLE
How to use extensions with Roboguice?

### ISSUE THREAD

**Issue by barnabycolby** [roboguice/roboguice#333]
I want to use throwing providers in my android project. This results in the com.google.inject.throwingproviders package not being found. How can I use this guice functionality in roboguice?

**Comment by stephanenicolas** [comment]
Did you add the guice dependency for throwing providers ?

2015-12-14 6:54 GMT-05:00 Barnaby Colby <notifications@github.com>:

**Comment by barnabycolby** [comment]
Yes, this results in the error "Multiple dex files define Lcom/google/inject/Module". I did manage to resolve this at some point only to find a runtime exception when instantiating my class using guice.

**Comment by stephanenicolas** [comment]
exclude the dependency in common when you import one of the 2 libs.

2015-12-14 8:33 GMT-05:00 Barnaby Colby <notifications@github.com>:

**Comment by barnabycolby** [comment]
Your suggestion got the compilation to succeed, although I'm not 100% sure that I excluded the right thing:
```java
compile 'org.roboguice:roboguice:3.+'
provided 'org.roboguice:roboblender:3.+'
compile('com.google.inject.extensions:guice-throwingproviders:3.+') {
    exclude module: 'guice'
}
```

This results in a runtime exception when the following line is executed:
```java
Guice.createInjector(new MyModule());
```

```java
java.lang.NoClassDefFoundError: Failed resolution of: Lcom/google/inject/internal/util/$Preconditions;
```

**Comment by stephanenicolas** [comment]
I got no idea about this one :(

Le mar. 15 déc. 2015 06:29, Barnaby Colby <notifications@github.com> a
écrit :

**Comment by barnabycolby** [comment]
Do you know if it's normally possible to use guice extensions with roboguice?

**Comment by rbaarda** [comment]
I'm having the same sort of problem with using the assisted inject extension.
This is my configuration:
```gradle
compile 'org.roboguice:roboguice:3.0.1'
provided 'org.roboguice:roboblender:3.0.1'
compile ('com.google.inject.extensions:guice-assistedinject:4.0') {
  exclude module: 'guice'
}
```

The exception I got is:
java.lang.NoClassDefFoundError: Failed resolution of: Lcom/google/inject/internal/util/$Maps;

I get the same exception when using assisted inject v3.

**Comment by stephanenicolas** [comment]
Not sure about RG 3, I think there is a bug open already for that.
But RG 4 that is actually already unofficially available on central drops
support for assisted injections for performance reasons.

RG 4 provides real performance gains but at the expense of more computation
intensive features in guice extensions. We wanted to push guice as far as
possible on Android.



2016-03-09 13:07 GMT-08:00 rbaarda <notifications@github.com>:

**Comment by rbaarda** [comment]
Thats good news, so when will RG4 be officially available?

I got one step further in using assisted inject and it might help @barnabycolby .
I completely removed my gradle cache, somehow jars for assisted-inject v3 and v4 got mixed up.
So after doing that, I've let gradle re-download all the dependencies and now assisted inject gave the same sort of exception, except this time it was looking for com/google/common/collect/Maps.
So I've added `compile 'com.google.guava:guava:19.0'` and this got rid of the exception.
Only now I'm stuck with 

> java.lang.IllegalStateException: Factories.create() factories cannot be used until they're initialized by Guice.

 When trying to use a factory in an Activity, but that is off topic for this issue.
